v.maxent.swd: bug fixes and new option to export explanatory raster layers

ecodiv opened the pull request:
- Bug fixes
- Added option to export the raster layers used as input in the creation of the SWD file. This can be used in Maxent as projection layers.

ecodiv committed:
v.maxent.swd: bug fixes and new option

- Bug fixes
- Added option to export the raster layers used as input in the creation of the SWD file. This can be used in Maxent as projection layers.

ecodiv committed:
Merge branch 'OSGeo:grass8' into v_maxent_swd

ninsbl commented on src/vector/v.maxent.swd/v.maxent.swd.html:
```diff
@@ -1,54 +1,80 @@
+which will be used by MaxEnt.
+
+<p>
+If alias names are used , a CSV file (alias_file) can be created with 
```
```suggestion
If alias names are used, a CSV file (alias_file) can be created with 
```

ninsbl commented on src/vector/v.maxent.swd/v.maxent.swd.html:
```diff
@@ -1,54 +1,80 @@
+(respecting region and mask, if set).
+
+<p>
+The SWD file format is a simple CSV-like file file format as described 
```
```suggestion
The SWD file format is a simple CSV-like file format as described 
```

ninsbl commented on src/vector/v.maxent.swd/v.maxent.swd.html:
```diff
@@ -1,54 +1,80 @@
+this addon you can have more than one sample point per raster cell. 
+points per raster cell. If this is not desired, use r.out.maxent_swd. 
+Another difference is the option to export the predictor raster layers 
+to an user-defined folder. This can be used in Maxent, Maxnet addon for 
```
```suggestion
to a user-defined folder. This can be used in Maxent, Maxnet addon for 
```

ninsbl commented on src/vector/v.maxent.swd/v.maxent.swd.py:
```diff
@@ -257,156 +289,213 @@ def main():
     else:
-        grass.run_command("g.copy", vector=[bgpn, bgpname], quiet=True)
-    grass.run_command("v.db.addcolumn", map=bgpname, columns=evp_cols, quiet=True)
+        if bool(options["seed"]):
+            gs.run_command(
+                "r.random",
+                input=environmental_layers[0],
+                npoints=bgpn,
+                vector=bgpname,
+                quiet=True,
+                seed=int(options["seed"]),
+            )
+        else:
+            gs.run_command(
+                "r.random",
+                input=environmental_layers[0],
+                npoints=bgpn,
+                vector=bgpname,
+                quiet=True,
+                flags="s",
+            )
```
You could simplify / shorten this:
```suggestion
        gs.run_command(
            "r.random",
            input=environmental_layers[0],
            npoints=bgpn,
            vector=bgpname,
            quiet=True,
            seed=int(options["seed"]) if options["seed"] else None,
        )
```

It may be enough (I guess) with:
```
seed=options["seed"],
```
But I have not tested...

ecodiv replied:
You're right, there is a default seed , so there is never a need to set the -s flag. Changing this.

ninsbl commented on src/vector/v.maxent.swd/v.maxent.swd.py:
```diff
@@ -257,156 +289,213 @@ def main():
-                    flags="c",
+    if bool(options["species"]):
+        gs.message(_("Creating SWD file with presence points"))
+        bgrdir = tempfile.mkdtemp()
```
You could use the GRASS-script tempfile functionality...

ecodiv replied:
Done, using grass.tempdir

ninsbl commented on src/vector/v.maxent.swd/v.maxent.swd.py:
```diff
@@ -257,156 +289,213 @@ def main():
+
+        # Get list with species names
+        specs = options["species"]
+        specs = specs.split(",")
```
```suggestion
        specs = options["species"].split(",")
```

ninsbl commented on src/vector/v.maxent.swd/v.maxent.swd.py:
```diff
@@ -257,156 +289,213 @@ def main():
+        if specsn == "":
+            specsn = [z.split("@")[0] for z in specs]
+        else:
+            specsn = specsn.split(",")
```
Maybe check if length of specsn is equal til length of specs?

ecodiv committed:
Merge branch 'OSGeo:grass8' into v_maxent_swd

ninsbl commented on src/vector/v.maxent.swd/v.maxent.swd.html:
```diff
@@ -1,54 +1,80 @@
 and subsequently used in other models / functions.
 
+<p>
+This addon was written as a Python alternative to the r.out.maxent_swd 
+script and provides much of the same functionality. The original script 
+has since been ported to Python (see also). One difference is that with 
```
Not sure if the history is so relevant for the manual. And there are good reasons for having this addon anyway. So I would drop the intro:
```suggestion
This addon is a vector based alternative to <em>r.out.maxent_swd</em>. It can
be more efficient with sparse data points. The main difference is that with
```

ninsbl commented on src/vector/v.maxent.swd/v.maxent.swd.py:
```diff
@@ -156,45 +182,68 @@
-    for rast in clean_rast:
-        grass.run_command("g.remove", type="rast", name=rast, quiet=True)
 
+def create_unique_name(name):
+    """Generate a tmp name which contains prefix
+    Store the name in the global list.
+    Use only for raster maps.
+    """
+    return name + str(uuid.uuid4().hex)
```
```
f"{name}{str(uuid.uuid4().hex)}"
```
should work, not sure if a function is needed here.

ninsbl commented on src/vector/v.maxent.swd/v.maxent.swd.py:
```diff
@@ -424,24 +513,20 @@ def main():
-        for fname in filenames:
-            with open(fname) as infile:
-                outfile.write(infile.read().rstrip() + "\n")
+        # Combine species swd files
+        filenames = os.path.join(bgrdir, "prespoints")
+        filenames = [filenames + str(i) for i in range(len(specs))]
+        with open(specout, "w") as outfile:
+            for fname in filenames:
+                with open(fname) as infile:
+                    outfile.write(infile.read().rstrip() + "\n")
```
Could`nt you just append? That would reduce I/O

ecodiv replied:
Not sure for now how best to implement that. Patching the point layers would be another way, but possibly slower. I am leaving it for now, but have put it on the list of possible improvements.

ninsbl replied:
I thought of opening output file in append mode:
https://stackoverflow.com/questions/4706499/how-do-i-append-to-a-file#4706520

ecodiv replied:
Ah, that simple :-). Learned (again) something new. Thanks.. changed

ninsbl commented on src/vector/v.maxent.swd/v.maxent.swd.html:
```diff
@@ -1,54 +1,80 @@
+If alias names are used , a CSV file (alias_file) can be created with 
+alias names in the first column and map names in the second column, 
+separated by comma, without header. This can be used as input in 
+r.maxent.lambdas (but note that this addon does not seem to work with 
+outputs from Maxent 3.4+).
```
Please open an issue / PR. From a brief look at the change log of Maxent I guess the issue is the transformation function!?

ecodiv replied:
@ninsbl I was planning, but first wanted to check if it is only the transformation. A quick test yielded a different result from Maxent using the logistic transformation. But I need to check that more carefully.

ninsbl reviewed (changes requested):
In principle this looks good!. I have not tested though.
Feel free to disregard suggestions where you would disagree.

ecodiv committed:
correct typo

Co-authored-by: Stefan Blumentrath <stefan.blumentrath@gmx.de>

ecodiv committed:
correct wording

Co-authored-by: Stefan Blumentrath <stefan.blumentrath@gmx.de>

ecodiv committed:
correct typo

Co-authored-by: Stefan Blumentrath <stefan.blumentrath@gmx.de>

ecodiv committed:
Update src/vector/v.maxent.swd/v.maxent.swd.py

Co-authored-by: Stefan Blumentrath <stefan.blumentrath@gmx.de>

ecodiv committed:
Update src/vector/v.maxent.swd/v.maxent.swd.html

Co-authored-by: Stefan Blumentrath <stefan.blumentrath@gmx.de>

ecodiv committed:
v.maxent.swd updates

Updates based on suggestions from @ninsbl

ecodiv committed:
extra check if species column exists

ecodiv committed:
use append instead of write to combine species files

ecodiv commented:
Hi @ninsbl, first, thanks for the suggestions. I followed up on all your suggestions, I think, but merging is still blocked. Do you know what else I need to do to go ahead?

echoix commented:
@wenzeslaus could we have automerge enabled here too? It could’ve been used here

neteler commented on src/vector/v.maxent.swd/v.maxent.swd.html:
```diff
@@ -1,54 +1,79 @@
+	<li>MaxEnt 3.4.1 (<a 
+	href="http://biodiversityinformatics.amnh.org/open_source/maxent/"> 
+	http://biodiversityinformatics.amnh.org/open_source/maxent/</a>)</li> 
+	<li><a 
+	href="https://grass.osgeo.org/grass-stable/manuals/addons/r.out.maxent_swd.html"> 
+	r.out.maxent_swd</a>, the original implemention using species 
+	distribution data in raster format.</li>
+	<li><a 
+	href="https://grass.osgeo.org/grass-stable/manuals/addons/r.maxent.lambdas.html"> 
+	r.maxent.lambdas</a> addon to computes raw or logistic prediction 
+	maps from MaxEnt lambdas files.</li>
+	
```
```suggestion
	<li><a href="r.out.maxent_swd.html">r.out.maxent_swd</a>, the original
	implementation using species distribution data in raster format.</li>
	<li><a href="r.maxent.lambdas.html">r.maxent.lambdas</a> addon to
	compute raw or logistic prediction maps from MaxEnt lambdas files.</li>
```

typo fixed + changed URLs
(absolute URLs only used for core modules while addon manual page URLs are relative)

ecodiv committed:
Update src/vector/v.maxent.swd/v.maxent.swd.html

Co-authored-by: Markus Neteler <neteler@osgeo.org>

ecodiv committed:
add test if env layers are exported

ecodiv committed:
remove link to script file

ecodiv committed:
Added check for square raster cells

Maxent only accepts raster cells with equal ns and ew resolution. Added code that checks the region setting, and gives a fatal message if ns resolution is not ew resolution. Optionally, the user can set the -e flag. This will automatically adjust the resolution of the largest side to match that of the smallest side.

ecodiv committed:
when exporting swd file, also export a proj. file with the crs info.

I plan to use this proj. file in r.maxent.train to check if the crs of the data used for the training matches that of the current mapset. If not, this proj. file will make it possible to import the output of r.maxent.train with reprojection (with warning). I can still have r.maxent.train issue a warning if there is no proj file, and let the user explicitly use the -o flag to override the projection check.

ecodiv committed:
Merge branch 'OSGeo:grass8' into v_maxent_swd

ecodiv committed:
Update v.maxent.swd.py

ecodiv committed:
Merge branch 'OSGeo:grass8' into v_maxent_swd
